### PR TITLE
Add a plugin system

### PR DESCRIPTION
**Pull request by toolness**
This adds a (hopefully) simple plugin system to allow NYCDB users to define their own custom datasets without needing to fork nycdb itself.

## Rationale

One use case for this is illustrated in [JustFixNYC/worst-evictors-data](https://github.com/JustFixNYC/worst-evictors-data), the data analysis done for the worst evictors site earlier this year.  The analysis used mostly NYCDB data, along with a small number of CSV files that were very specific to the analysis.  Nonetheless, these files contained "NYCDB-esque" data and it felt easiest to think of them as additional datasets that we wanted to use NYCDB to extract, transform, and load (rather than e.g. creating a separate set of Python scripts that duplicated so much of NYCDB's built-in ETL functionality).

I ended up implementing a terrible hack in [`monkeypatch_nycdb/patch.py`](https://github.com/JustFixNYC/worst-evictors-data/blob/master/monkeypatch_nycdb/patch.py) which essentially "tricked" NYCDB into thinking it had extra datasets.  However, rather than requiring such tomfoolery, I thought it would be nice if NYCDB had some kind of simple extensibility mechanism which might make this easier.

## Plugin structure/API

An NYCDB plugin is just a Python package which mimicks a subset of the `nycdb` package's own structure:

* A `dataset_transformations.py` file contains transformations for datasets the plugin defines, using the exact same naming/calling convention as `nycdb.dataset_transformations`.

* A `datasets` folder contains YAML files defining datasets and their metadata.

* A `sql` folder contains optional SQL transformations referenced by dataset YAML files.

All of this should look really familiar; in fact, because `nycdb` itself has this very structure, we can simply consider it just another plugin, which simplifies implementation considerably (i.e., there's no need for NYCDB's source code to distinguish between the datasets provided by `nycdb` and those provided by plugins).

## Plugin discovery

A third-party NYCDB plugin's package name should start with `nycdb_`.  This follows the [naming convention](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-naming-convention) strategy for plugin discovery, also used by Flask, which makes enabling plugins as straightforward as `pip install`'ing them, or otherwise making sure they're in `sys.path`.

In fact, the plugin integration test included in this PR works by temporarily adding an example plugin's parent directory to `sys.path` and then removing it once the test is over.

## Example

An example plugin is located in this PR at `src/tests/integration/plugins/nycdb_myplugin`:

```
.
└── nycdb_myplugin
    ├── __init__.py
    ├── dataset_transformations.py
    ├── datasets
    │   └── myplugin_custom_dataset.yml
    └── sql
        └── myplugin_custom_dataset.sql
```

This plugin defines a dataset called `myplugin_custom_dataset`.  A sample CSV for this fake dataset is in `src/tests/integration/data/myplugin_custom_dataset.csv`.

## Other benefits

Aside from the use case mentioned in the rationale section above, there are other potential benefits to adding this extensibility mechanism to NYCDB.

For instance, because it allows people to add datasets to NYCDB without needing to either fork it or request permission for inclusion into the next release of NYCDB, potential new datasets can easily be created and iterated on independently.  People can easily "try out" a third-party dataset on their machines via `pip install` and, if a lot of folks find it really useful and its schema has had some time to settle, it can be nominated for inclusion into the next release of NYCDB.

## Risks

* As with any extensibility system, plugins might break with new releases of NYCDB.  This could happen, for instance, if the location or call signature of a utility function imported in a plugin's `dataset_transformations` changes, or if the YAML structure describing a dataset changes in a backwards-incompatible way.  We'd need to be aware of such potential breakages and ideally first start by deprecating a backwards-incompatible change (e.g. logging a [`DeprecationWarning`](https://docs.python.org/3/library/exceptions.html#DeprecationWarning) whenever a deprecated feature is accessed) and only fully removing it in a future release.

* By lowering the barrier for third parties to extend NYCDB, we might end up with a bunch of third-party datasets that are hard for community members to discover, rather than having them be in NYCDB core and thus be easy to find.  I've tried to make it as easy as possible to uplift third-party plugins into NYCDB itself, though: because the file structure is identical, the process should just consist of copying the relevant YAML and SQL files, and merging the plugin's `dataset_transformations.py` into `nycdb.dataset_transformations`.  Er, and adding some tests.  And adding the validation thingy.

## To do

If this proposal is accepted, there's a few things I'd like to add, either as follow-up PRs, or as commits before merging this one:

* I'd like to formally document the plugin system, probably by moving a lot of this PR's description into a markdown file.

* Some of the docstrings in functions I changed may now be inaccurate; I should double-check them and make sure they're correct.

* I'd like to add tests for some of the edge cases, such as when an SQL file or dataset transformation function can't be found.

### CODE REVIEW

**Comment by aepyornis** [comment]
I need to spend a little bit of time thinking about this!

but as a solution to your immediate problem -- is there a reason why we can't add the worst-evictors-data to nycdb outright?

**Comment by toolness** [comment]
That's not my immediate problem; I've wanted this kind of functionality for a long time, in a variety of other use cases, and the rationale I mentioned was just one example.

Others include e.g.:

* Who Owns What builds on NYCDB but we have a [big script](https://github.com/JustFixNYC/who-owns-what/blob/master/dbtool.py) that essentially just runs a bunch of SQL after installing some datasets.  If we could leverage more of nycdb itself, e.g. by considering WoW to just be a dataset that has other datasets as dependencies, this could be avoided.  (This would require building in the concept of dataset dependencies to NYCDB, but it's something we're likely going to need to do anyways due to how we'll likely be evolving the tax bills dataset, but that's a separate conversation).

* Adding the worst evictors data to NYCDB outright would have required thinking about backwards compatibility and consulting with the community, but the priority was really to get the worst evictors study done, not to create a dataset that would stand the test of time.  Being able to leverage nycdb to create a "draft dataset" in our own repository would decouple the need to get our work done from the process of adding a mature dataset to a community resource.  For example, while we were working on our data analysis, we could talk to the community about how a more permanent NYCDB dataset might look; we could then evolve our own dataset to look like it, while still working on our data analysis.

* We get a lot of data requests at JustFix that involve queries between NYCDB data and other arbitrary datasets.  Right now @sraby is the main person who works on these, and he uses [Postico](https://eggerapps.at/postico/) to import CSVs into his local Postgres database (which also has NYCDB on it) and then make queries.  However, this isn't very reproducible, and Postico isn't even available for non-Mac systems.

  As far as I know, there's no easy tool to help me dump a CSV potentially needing NYC-relevant data transforms into a Postgres database in a reproducible way.  Maybe I'm just using nycdb for the wrong thing?  But writing a Python script from scratch to do it is also a lot of work and I feel like it's easier to just leverage NYCDB's existing plumbing by telling it that I have a new dataset.  That said, I'm also new to the ETL world so maybe I'm missing something.

File: src/nycdb/plugins.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        self.name = name
+        self.spec = find_spec(name)
+        self.root_dir = Path(self.spec.origin).parent.resolve()
+
+    def import_submodule(self, name: str):
+        return importlib.import_module(f"{self.name}.{name}")
+
+
+def iter_plugins():
+    yield Plugin('nycdb')
```
Comment (aepyornis) [inline src/nycdb/plugins.py]: iIf nycdb is a "plugin" than perhaps "plugin" is the right name for this then? component?

File: src/nycdb/dataset.py [hunk 2]
```diff
@@ -64,9 +65,32 @@ def index(self):
...
+    def _get_transformed_dataset_rows(self, schema):
+        for plugin in plugins.iter_plugins():
+            dt_module = plugin.import_submodule('dataset_transformations')
+            rows = self._get_transformed_dataset_rows_using_module(schema, dt_module)
+            if rows is not None:
+                return rows
+
+        table_name = schema['table_name']
+        raise Exception(
+            f"Unable to find transformer for dataset {self.name}, "
```
Comment (aepyornis) [inline src/nycdb/dataset.py]: I like this name. We think we should rename the module `dataset_transformations` to `transformers`

File: src/nycdb/dataset.py [hunk 3]
```diff
@@ -64,9 +65,32 @@ def index(self):
...
+    def _get_transformed_dataset_rows_using_module(self, schema, dataset_transformations_module):
+        rows_func = getattr(dataset_transformations_module, schema['table_name'], None)
+        if rows_func:
+            return rows_func(self)
+        rows_func = getattr(dataset_transformations_module, self.name, None)
+        if rows_func:
+            return rows_func(self, schema)
+        return None
+
+    def _get_transformed_dataset_rows(self, schema):
```
Comment (aepyornis) [inline src/nycdb/dataset.py]: I find this section confusing. and, this is not to say what we had before was good -- that was also confusing!

i think we should deal with transformations differently. Perhaps we could try something like having a set of common transformations and then link them together in dataset yaml file:
``` yaml
     transformations: 
         - csv
         - bbl 
``` 

regardless of that proposal, I think we should refactor the transformer/transformations code (a lot of the confusion was my fault when I introduced the option of using a transformer named for either the dataset or the schema name)

File: src/nycdb/plugins.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    for finder, name, ispkg in pkgutil.iter_modules():
+        if ispkg and name.startswith('nycdb_'):
```
Comment (aepyornis) [inline src/nycdb/plugins.py]: take this comment with these caveats -- i'm not  a python packaging expert or a security expert.

does give open up some security concerns? What's stopping someone from creating package called nycdb_evil_plugin, uploading it to pypi and having it get installed automatically here?

Comment (toolness) [reply src/nycdb/plugins.py]: Oh, this will only activate the plugin if it's locally installed on the user's system, not if it's available on PyPI.  As with any package (since packages can execute arbitrary code when they're installed), a developer is expected to make sure they're not installing anything malicious when they `pip install` it.  But once a package is on a user's system it's assumed to not be malicious.

Also, given that this [convention](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-naming-convention) is recommended by the official Python documentation and used by Flask (a package that takes security very seriously) I think we should be good!

File: src/nycdb/dataset.py (hunk 3 above)
Comment (toolness) [reply src/nycdb/dataset.py]: Ah good idea, I did find this code confusing to write, so I think it'd be nice to be more explicit with the transformations if possible!

File: src/nycdb/dataset.py (hunk 2 above)
Comment (toolness) [reply src/nycdb/dataset.py]: Ah nice, let's do that!

File: src/nycdb/plugins.py (hunk 1 above)
Comment (toolness) [reply src/nycdb/plugins.py]: Hmmm, I'm concerned that `Component` might also be a bit overly generic though?  I also feel like the word `plugin` has a certain understanding within the Python community (see e.g. the [Creating and discovering plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/) docs on the official Python site) that conforms to what we're doing here, so I'd prefer to keep using that word if possible.

Another option is for us to actually have nycdb come with a separate package called `nycdb_core` that represents the core "built-in" plugin but I'm not sure if/how much that would help.

File: src/nycdb/dataset.py (hunk 3 above)
Comment (toolness) [reply src/nycdb/dataset.py]: Hmm, so it seems like we should probably hold off on merging this PR until this whole transformations API is changed.  Do you want to write up a PR for that or should I?

**Comment by aepyornis** [comment]
closing this since it's a been while, we can always revisit later if there is interest.
